frontend/src/Components/ContentComponent.tsx

```diff
@@ -56,6 +78,123 @@ function updateContent(div: HTMLDivElement, setZoomedImg: (img: ZoomedImg | null
     div.querySelectorAll('details.expand').forEach(expand => {
         updateExpand(expand as HTMLDetailsElement);
     });
+
+    div.querySelectorAll('span.secret-mailbox').forEach(mailbox => {
+        updateMailbox(mailbox as HTMLSpanElement, setMailboxKey);
+    });
+
+    div.querySelectorAll('span.secret-mail').forEach(mail => {
+        updateMail(mail as HTMLSpanElement, setMailboxKey, currentUsername);
+    });
+}
+
+function updateMailbox(mailbox: HTMLSpanElement,
+                          setMailboxKey: (key: MailboxKey | null) => void
+                       ) {
+    const secret = mailbox.dataset.secret;
+    if (!secret) {
+        return;
+    }
+    let mailboxTitle = mailbox.dataset.rawText;
+    try {
+        mailboxTitle = mailboxTitle && b64DecodeUnicode(mailboxTitle);
+    } catch (e) {
+        mailboxTitle = undefined;
+    }
+
+    mailbox.addEventListener('click', () => {
+        setMailboxKey({
+            type: 'mailbox',
+            openKey: secret,
+            mailboxTitle
+        });
+    });
+}
+
+function updateMail(mail: HTMLSpanElement,
+                    setMailboxKey: (key: MailKey | null) => void,
+                    currentUsername?: string
+) {
+    // check processed
+    if (mail.dataset.processed) {
+        return;
+    }
+    mail.dataset.processed = '1';
+
+    const secret = mail.dataset.secret;
+    if (!secret) {
+        mail.classList.add('secret-mail-error');
+        return;
+    }
+    let cipher: string | undefined;
+    let encodedKey: string;
+
+    // try decode secret as json
+    try {
+        const j = JSON.parse(b64DecodeUnicode(secret));
+        // add mention "для @username"
+        if (j.to && !mail.querySelector('span.mention')) {
+            const mention = document.createElement('span');
+            mention.classList.add('mention');
+            mention.innerText = j.to;
+            mail.appendChild(document.createTextNode(' для '));
+            mail.appendChild(mention);
+        }
+
+        // check v, to, c
+        if (!j.v || !j.c || !Number.isInteger(j.v) || !j.toKey) {
+            throw new Error('Invalid secret');
+        } else if (j.to && j.to === currentUsername && j.toKey) {
+            encodedKey = j.toKey;
+            cipher = j.c;
+        } else if (j.from && j.from === currentUsername && j.fromKey) {
+            encodedKey = j.fromKey;
+            cipher = j.c;
+        } else if (!j.to) {
+            encodedKey = j.toKey;
+            cipher = j.c;
+        }
```

Comment:
it is possible to modify this code on the client side and start guessing other's people's passwords. Ideally we need a check on the backend, that only sender or recipient are trying to decipher

Reply:
Doesn't really matter. All the encoded info is in the open — anybody can try to bruteforce. No protection against it, unless we move all encoding to the backend, but that's kinda against the point — to have encryption transparent and not send anything to the server.